### DISCUSSION TITLE
RabbitMQ Single Node Latency / Memory / Throughput limits

### DISCUSSION THREAD

**Discussion by JanEggers** [rabbitmq/rabbitmq-server#8248] [Questions]
we are currently using RabbitMq single node 3.11 deployment to communicate with our PLCS. The PLCS use MQTT and we use a controller type app that uses amqp that subscribes to all events from the plcs and sends commands to controll them.

We observed some things where we are not sure whats the reason / if there are ways we can improve.

Throughput:
We use MQTT explorer to monitor the system. It subscribes to all messages. 1 running Instance of it fine 1200 msg/sec. But with 2 running instances  the outbound traffic reduces to 600 msg/ sec and we see items buildup in the queues. This happens at network utilization of up to 30 mb / sec. But we run GB so there should be room for 5x more traffic without queue buildup. (CPU is at 20%)

are there any limits to rabbit that limit the traffic rate besides network constraints?

Memory usage:

the brokers memory usage increases up to several GB although all Queues are empty. the memory is accumulated in binary references whatever that means (diagnosed with admin site). and after a reboot of the broker we are back to 200 MB memory usage.

Latency:
we added logging to our app that shows roundtrip latency. the pattern is always PLCS sends event. controller sends command. PLC loggs latency bewteen event and command. everything is running in the same local network but the roundtrip times are in a wild range from 40ms to 2 sec. 

is this normal? is there any way to improve latency?

**Comment by lukebakken** [comment]
Please try a 3.12 release candidate, there are *huge* improvements to RabbitMQ's MQTT support -

https://blog.rabbitmq.com/posts/2023/05/rabbitmq-3.12-performance-improvements/#new-mqtt-implementation-significant-memory-savings-per-connection-supports-millions-of-connections-per-node

cc @ansd

**Comment by JanEggers** [comment]
just did that. we had to roll back as connections have been shaky: https://github.com/rabbitmq/rabbitmq-server/issues/8251

**Reply by JanEggers** [reply]
we are now running 3.12-rc3 but latency (our biggest pain point) did not really improve. 

is there any way to generate performance related reports that indicate why we see latency issues?

**Reply by ansd** [reply]
1. Did memory usage decrease with 3.12-rc3? If so, by how much?
2. Did throughput increase with 3.12-rc3? If so, by how much?
3.  Based on your current description, there is very little we can anlayze right now. Please provide reproducible steps if you want us to analyse latency issues. Best would be if you could provide a GitHub repo such as https://github.com/ansd/rabbitmq-mqtt which deploys RabbitMQ including MQTT clients. Since you experience latency issues, we'll need to know all variables: how many MQTT clients? What QoS level? What is the message size? How many messages per minute does each MQTT client send? How many messages per minute does the AMQP 0.9.1 client send? From what you describe, MQTT devices do a fan-in topology to the AMQP 0.9.1 client. How does the AMQP 0.9.1 client publish messages to the MQTT devices (1:1 or fan-out)? What resources did you deploy RabbitMQ with (CPU, memory, disks (SSD?))?

> we see items buildup in the queues

Which queues exactly?

> are there any limits to rabbit that limit the traffic rate besides network constraints?

Yes, for example flow control to protect the broker from being overloaded: https://www.rabbitmq.com/flow-control.html


> is this normal

Hard to say with the information you provided

> is there any way to improve latency?

Possibly, yes:
* Disable management_agent metrics and use classic queues v2 (see https://github.com/ansd/rabbitmq-mqtt/blob/3d4ddf83ca00a2cbbe15e49b307a8c989c2dcc8f/one-to-one/mqtt-rabbit.yml#L25-L26 on how to configure this)
* Make sure you have deployed RabbitMQ with sufficient CPUs
* Use QoS 0 instead of QoS 1 if you want lower latency.

> is there any way to generate performance related reports that indicate why we see latency issues?

If RabbitMQ is maxed out on CPU usage, you could provide us flame graphs as described in https://blog.rabbitmq.com/posts/2022/05/flame-graphs

**Reply by pfeifi01** [reply]
Some questions are hard to answer, because it's not easy to observe, but i'll try it to my best knowledge:

1. It may be that the memory usage decreased when looking at the current RAM usage in the Task Manager. 

![image](https://github.com/rabbitmq/rabbitmq-server/assets/17936009/3e21f948-18a2-4a31-b024-ac1d037f26f9)

But the problem we had regarding the RAM, is that we were accumulating huge load of binaries over time, that were never reset and went back to zero, after we restarted the RabbitMQ service. So to come to a conclusion if this has increased, it may take days to reproduce the behavior.

Here is an example of how the system was behaving with the rabbit version 3.11, 

Here we can see the RAM accumulation:

![MicrosoftTeams-image (47)](https://github.com/rabbitmq/rabbitmq-server/assets/17936009/be806cd5-7906-4192-8596-f7f0391483ab)

And here we can see the reset of the RAM after the service was restarted:

![MicrosoftTeams-image (48)](https://github.com/rabbitmq/rabbitmq-server/assets/17936009/ef7bbbcb-ecd4-44f6-8058-6fbac74b5987)

2. To me it seems like throughput increased, but i can also not really undermine this with data. Generally when checking the Task Manager it seems like we are now able to get more data out through the network than before, but this can also just be a coincidence. 

We still see huge delays in communication and roundtriptimes some times for events. For each of our PLC-MQTT-Clients we have a "heartbeat" subscription, which is an event/command that is being sent and acknowledged every second, and this regularly needs over 1 second for an answer, if we are pressuring the system a lot, this delay can increase up to 10 seconds.

![image](https://github.com/rabbitmq/rabbitmq-server/assets/17936009/ea66b3dc-6675-483d-b3a6-577fc8637401)

We are using the MQTT Explorer to observe and manually send events from time to time, and we can see that when we are using this tool, that we have a huge number of messages "_ready_", which we are not able to deliver in a timely fashion through the network and for each MQTT explorer increases the ready-messages hugely.

3. 
I'll provide you the data i'm able to collect

- We currently have 49 connections that all looks kind of like this, where most are using the MQTT protocol:

![image](https://github.com/rabbitmq/rabbitmq-server/assets/17936009/0e4a2c03-b3c7-4efc-aee3-a29840e7f869)

- We have 10 channels:

![image](https://github.com/rabbitmq/rabbitmq-server/assets/17936009/d3955760-212f-46ba-a53b-740ef65aaecb)

- We have around 150 Queues:

![image](https://github.com/rabbitmq/rabbitmq-server/assets/17936009/559f1adf-54fb-417f-869d-f0b474af7029)

and when we open an MQTT explorer, it's getting added to the queue and we can see messages accumulating in the "Ready" cell and at some point being reset. But if the system experiences a heavier message load, we have observed that the value in the "Ready" cell is just infinitely increasing and cannot be processed to zero by the broker.

- Quos is 0
- message size is 100 - 150 bytes 
- amqp client publishes 1:1

### Hardware
- we are running this on a Windows Server shared VM environment, where we have (Microsoft Windows Server 2019)

**Hardware Specs:**
- SSD 80GB Total, 45GB free
- RAM (32GB available)
- Intel(R) Xenon(R) Gold 6154 @3.00 GHz (8 Cores available)

RabbitMQ is installed on C:\
![image](https://github.com/rabbitmq/rabbitmq-server/assets/17936009/eb276a6c-8526-4ebb-9005-3f9de42b52a8)

- we are not maxed out in RAM, CPU or Network, but we seem not to be able to exchange messages without delay with the current load on the system, which should not be that much, as we anticipate more load in the future.

i hope this helps

**Reply by ansd** [reply]
Thanks for the infos and screenshots.

So, you use MQTT + AMQP 0.9.1 + STOMP (from what I see in the screenshots).

Let's fully focus on 3.12 and forget 3.11 for now.

Just to make sure I got it right: [This](https://user-images.githubusercontent.com/17936009/240199423-ea66b3dc-6675-483d-b3a6-577fc8637401.png) screenshot is using 3.12 with [feature flag](https://www.rabbitmq.com/feature-flags.html)  [rabbit_mqtt_qos0_queue](https://blog.rabbitmq.com/posts/2023/03/native-mqtt/#new-mqtt-qos-0-queue-type) disabled, correct?
I assume so, because the MQTT queues show `D` for `durable` and `Excl` for `exclusive`, the queue names end with `qos0`, and queue types are `classic`.

As general guideline, please ensure **all** feature flags are enabled in 3.12!
Also make sure you use [classic queues version 2](https://rabbitmq.com/persistence-conf.html#queue-version).

Given messages accumulate in the queues of MQTT explorer, it means that messages arrive faster into these queues than can be delivered to the MQTT explorer clients. We see that 567 messages/second arrive into these queues.

With feature flag `rabbit_mqtt_qos0_queue` enabled, messages will be dropped if the MQTT client cannot consume fast enough.

With that feature flag disabled, you could also try increasing `mqtt.prefetch` (see https://www.rabbitmq.com/mqtt.html#stickiness):
> The prefetch option controls the maximum number of unacknowledged messages that will be delivered. This option is interpreted in the same way as the [AMQP 0-9-1 prefetch-count](https://www.rabbitmq.com/amqp-0-9-1-reference.html#basic.qos.prefetch-count) field, so a value of 0 means "no limit".

The default prefetch of `10` is likely too low in your scenario where the queue receives more than 500 messages per second. However, I don't think that's the issue we're seeing here because the MQTT connection process consumes in [automatic ack mode](https://github.com/rabbitmq/rabbitmq-server/blob/f82a3f253deb818a685906fba0ce2f17be919a21/deps/rabbitmq_mqtt/src/rabbit_mqtt_processor.erl#L993).

You need to make sure that the MQTT explorer client apps are able to consume messages fast enough.

**Reply by ansd** [reply]
FWIW, I downloaded MQTT Explorer version `0.3.5`, started it on the MAC and subscribed to topic `#`.
I started RabbitMQ 3.12.0-rc.3 as follows
```
make run-broker RABBITMQ_CONFIG_FILE="rabbitmq.conf" FULL=1 PLUGINS="rabbitmq_management rabbitmq_mqtt" RABBITMQ_FEATURE_FLAGS=virtual_host_metadata,user_limits,quorum_queue,maintenance_mode_status,implicit_default_bindings,classic_mirrored_queue_version
```
where `rabbitmq.conf` contains:
```
classic_queue.default_version = 2
```
This leaves feature flag `rabbit_mqtt_qos0_queue` disabled (as in your environment).

I then generate load on the same local machine (MAC): 100 MQTT clients publish to topic `a/b` each of them every 20ms, therefore 100 * 50 = 5,000 messages per second. Each message has a size of 150 bytes.
```
./emqtt_bench pub --version 4 --count 100 -t a/b --interval_of_msg 20 --size 150
```
MQTT explorer was able to consume these messages without any issues.

**Reply by ansd** [reply]
When I send with much higher rates I see temporarily that ~1k messages queue up in the MQTT Explorer classic queue and that consumer utilisation is around 50%. However, eventually the queue is being emptied. So, it all depends on how fast MQTT explorer can consume the messages.

Long story short is: Right now I don't see any performance issues with RabbitMQ itself.
Make sure your MQTT clients can consume fast enough - if they don't: TCP backpressure is applied to the server preventing the server from sending fast enough resulting in messages being accumulated in the classic queue leading to high end-to-end latency.

**Reply by JanEggers** [reply]
thanks a lot for your support / advice.

we will try classic_queue.default_version = 2.

rabbit_mqtt_qos0_queue is not viable for us as we should not drop messages

just to clarify:

- if messages are accumulated in one queue because the consumer is not fast enough. does that increase the latency for other queues / cllients.

- we thought about switching to a multi node / cluster configuration to distribute load / reduce latency.
is this benificial?

**Reply by ansd** [reply]
> rabbit_mqtt_qos0_queue is not viable for us as we should not drop messages

If you must not drop messages, you should use QoS 1.
If you should not drop messages, you could also increase `mqtt.mailbox_soft_limit` to some higher value (than the current default of 200).

> if messages are accumulated in one queue because the consumer is not fast enough. does that increase the latency for other queues / cllients.

No.

> we thought about switching to a multi node / cluster configuration to distribute load / reduce latency.
is this benificial?

In general, it could. But I don't think this will solve the issue you are experiencing here: That the MQTT explorer client cannot consume the messages fast enough.

**Reply by michaelklishin** [reply]
There is a claim made above about nodes "accumulating binaries", with a screenshot that clearly demonstrates that what consumes seemingly well over 90% of memory is classic queues.

They can contain unconsumed messages or not, but they are not binary heap fragments.

Most likely what happens with this workload is that consumers come and go, and leave queues behind. RabbitMQ will not delete them unless those queues are exclusive (not necessarily a good idea) or have TTL configured for them (using a policy, MQTTv3 clients do not have a way of doing that).

So setting a queue TTL of 3-12 hours should help a lot. Note that queues that have consumers will not be deleted due to TTL.

**Reply by michaelklishin** [reply]
The MQTT client in 3.11, and likely comparably in 3.12, can do 5-10K messages a second *per connection*. There is a realistic limit to how far that can go with more connections but double digit thousands sounds perfectly achievable with efficient clients.

Network bandwidth is not the only limiting factors. Disk I/O and CPU cores should not be ignored. RabbitMQ will happily use more cores if there are more connections/queues (subscribers in case of MQTT).

You can always try PerfTest and see what your setup is capable of. Yes, it does not use MQTT but it will establish a baseline. PerfTest is a mature, very powerful tool that can simulate a broad range of workloads.

We do not have an equivalent for MQTT but there should be a similar tool in the MQTT community. Use that instead of guessing, and collect metrics in the process.

**Comment by pfeifi01** [comment]
We have the old format of the config file and we were trying to adjust it to set the classic queue version to 2. We are not perfectly sure if we have adjusted it correctly. Is there any way to check via the RabbitMQ Web Dashboard if the queues have now version 2?

**Reply by ansd** [reply]
There is in 3.12.0-rc.3 thanks to @gomoripeti : https://github.com/rabbitmq/rabbitmq-server/pull/8241
In the Management UI, click on the classic queue details, there should be a field called `queue storage version`.

**Reply by michaelklishin** [reply]
We should expose this via CLI tools, too. I will file an issue.

**Comment by JanEggers** [comment]
can you please have a look at this repro: 
https://github.com/JanEggers/RabbitMqLatencyBenchmark

the more producers there are the worse the latency. i reduced the numbers of messages send the more clients i used so the message rate is constant.

also it seems to be a general issue regardless of the protocol when I use amqp as producer i can also see degrading latency the more producers there are

**Reply by JanEggers** [reply]
i finally was able to configure queue version 2 but that just brought minor improvements.

**Reply by JanEggers** [reply]
i updated the benchmark again. I added a dotnet mqtt broker to funnel all messages through a single amqp connection. that way i was able to bring down latency to 30ms and improve throughput by 3x.

im a c# dev so im totally useless helping you to figure out what causes the bottleneck in rabbit because i dont know erlang. but there seems to be some kind of contention if the number of publishers increase.
